File: frontend/src/api/student/reports.ts
import { apiClient } from '../apiClient';

export interface DetailedReport {
  id: number;
  user_id: number;
  report_type: string;
  title: string;
  description?: string;
  period_start: string;
  period_end: string;
  data: any;
  insights?: string;
  recommendations?: any;
  is_exported: boolean;
  exported_at?: string;
  created_at: string;
}

export interface SubjectProgressReport {
  id: number;
  user_id: number;
  subject: string;
  period_start: string;
  period_end: string;
  total_score: number;
  max_score: number;
  percentage: number;
  improvement_rate?: number;
  topics_covered?: any;
  strengths?: any;
  weaknesses?: any;
  recommendations?: any;
  created_at: string;
}

export interface AnalyticsReport {
  id: number;
  user_id: number;
  analytics_type: string;
  period_start: string;
  period_end: string;
  metrics: any;
  trends?: any;
  insights?: string;
  recommendations?: any;
  created_at: string;
}

export interface ReportExport {
  id: number;
  report_id: number;
  export_format: string;
  file_url?: string;
  file_size?: number;
  exported_at: string;
  status: string;
}

export interface CreateDetailedReportData {
  report_type: string;
  title: string;
  description?: string;
  period_start: string;
  period_end: string;
}

export interface CreateSubjectProgressReportData {
  subject: string;
  period_start: string;
  period_end: string;
}

export interface CreateAnalyticsReportData {
  analytics_type: string;
  period_start: string;
  period_end: string;
}

class ReportsAPI {
  // Rapports détaillés
  async getDetailedReports(report_type?: string, limit?: number): Promise<DetailedReport[]> {
    const params = new URLSearchParams();
    if (report_type) params.append('report_type', report_type);
    if (limit) params.append('limit', limit.toString());
    
    return this.request(`/api/v1/reports/detailed?${params.toString()}`);
  }

  async getDetailedReport(reportId: number): Promise<DetailedReport> {
    return this.request(`/api/v1/reports/detailed/${reportId}`);
  }

  async createDetailedReport(data: CreateDetailedReportData): Promise<DetailedReport> {
    return this.request('/api/v1/reports/detailed', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // Rapports de progression par matière
  async getSubjectProgressReports(subject?: string, limit?: number): Promise<SubjectProgressReport[]> {
    const params = new URLSearchParams();
    if (subject) params.append('subject', subject);
    if (limit) params.append('limit', limit.toString());
    
    return this.request(`/api/v1/reports/subject-progress?${params.toString()}`);
  }

  async createSubjectProgressReport(data: CreateSubjectProgressReportData): Promise<SubjectProgressReport> {
    return this.request('/api/v1/reports/subject-progress', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // Rapports d'analytics
  async getAnalyticsReports(analytics_type?: string, limit?: number): Promise<AnalyticsReport[]> {
    const params = new URLSearchParams();
    if (analytics_type) params.append('analytics_type', analytics_type);
    if (limit) params.append('limit', limit.toString());
    
    return this.request(`/api/v1/reports/analytics?${params.toString()}`);
  }

  async createAnalyticsReport(data: CreateAnalyticsReportData): Promise<AnalyticsReport> {
    return this.request('/api/v1/reports/analytics', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // Export de rapports
  async exportReport(reportId: number, export_format?: string): Promise<ReportExport> {
    const params = new URLSearchParams();
    if (export_format) params.append('export_format', export_format);
    
    return this.request(`/api/v1/reports/${reportId}/export?${params.toString()}`, {
      method: 'POST',
    });
  }

  private async request<T>(endpoint: string, options?: RequestInit): Promise<T> {
    const response = await apiClient.request<T>(endpoint, options);
    return response.data;
  }
}

export const reportsAPI = new ReportsAPI();

File: backend/models/reports.py
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, ForeignKey, Float, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from core.database import Base
from datetime import datetime

class DetailedReport(Base):
    __tablename__ = "detailed_reports"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    report_type = Column(String(50), nullable=False)  # performance, progress, analytics, behavior
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    period_start = Column(DateTime(timezone=True), nullable=False)
    period_end = Column(DateTime(timezone=True), nullable=False)
    data = Column(JSON, nullable=False)  # Données du rapport
    insights = Column(Text, nullable=True)  # Insights générés
    recommendations = Column(JSON, nullable=True)  # Recommandations
    is_exported = Column(Boolean, default=False)
    exported_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relations
    user = relationship("User", foreign_keys=[user_id])

class SubjectProgressReport(Base):
    __tablename__ = "subject_progress_reports"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    subject = Column(String(100), nullable=False)
    period_start = Column(DateTime(timezone=True), nullable=False)
    period_end = Column(DateTime(timezone=True), nullable=False)
    total_score = Column(Float, nullable=False)
    max_score = Column(Float, nullable=False)
    percentage = Column(Float, nullable=False)
    improvement_rate = Column(Float, nullable=True)
    topics_covered = Column(JSON, nullable=True)  # Liste des sujets couverts
    strengths = Column(JSON, nullable=True)  # Points forts
    weaknesses = Column(JSON, nullable=True)  # Points faibles
    recommendations = Column(JSON, nullable=True)  # Recommandations
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relations
    user = relationship("User", foreign_keys=[user_id])

class AnalyticsReport(Base):
    __tablename__ = "analytics_reports"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    analytics_type = Column(String(50), nullable=False)  # engagement, performance, behavior, learning_patterns
    period_start = Column(DateTime(timezone=True), nullable=False)
    period_end = Column(DateTime(timezone=True), nullable=False)
    metrics = Column(JSON, nullable=False)  # Métriques calculées
    trends = Column(JSON, nullable=True)  # Tendances identifiées
    insights = Column(Text, nullable=True)  # Insights générés
    recommendations = Column(JSON, nullable=True)  # Recommandations
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relations
    user = relationship("User", foreign_keys=[user_id])

class ReportExport(Base):
    __tablename__ = "report_exports"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    report_id = Column(Integer, ForeignKey("detailed_reports.id"), nullable=False)
    export_format = Column(String(20), nullable=False)  # pdf, excel, csv, json
    file_url = Column(String(500), nullable=True)
    file_size = Column(Integer, nullable=True)  # en bytes
    exported_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relations
    user = relationship("User", foreign_keys=[user_id])
    report = relationship("DetailedReport")
